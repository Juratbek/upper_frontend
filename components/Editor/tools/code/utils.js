export function renderSettings(settings, onClick, context) {
  const wrapper = document.createElement("div");

  settings.forEach((tune) => {
    const button = document.createElement("div");

    button.classList.add(context.api.styles.settingsButton);
    button.classList.add(tune.className);
    button.onclick = () => {
      onClick(tune);
      button.classList.toggle(context.api.styles.settingsButtonActive);
    };
    button.innerHTML = tune.icon;
    wrapper.appendChild(button);
  });

  return wrapper;
}
