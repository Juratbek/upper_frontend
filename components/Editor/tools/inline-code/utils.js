export const isTextEmpty = (textContent) => {
  return /^\s*$/.test(textContent);
};

export const isCaretAtLastPosition = (element) => {
  // Get the selection object
  const selection = window.getSelection();

  if (selection.rangeCount === 0) return false;

  // Get the range object
  const range = selection.getRangeAt(0);

  // Check if the end offset of the range is equal to the length of the text content
  return range.endOffset === element.textContent.length;
};

export const isTextNode = (node) => {
  return node && node.nodeType === 3;
};

export const startsWithCharacters = (text) => {
  if (typeof text !== 'string') return false;
  return /^\S/.test(text);
};
