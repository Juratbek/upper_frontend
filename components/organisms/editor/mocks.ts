import { vi } from 'vitest';

import { IEditorAPI } from './context';

export const mockApi: IEditorAPI = {
  addBlock: vi.fn(),
  addBlocks: vi.fn(),
  hideInlineToolbar: vi.fn(),
  mergeWithPrevBlock: vi.fn(),
  moveBlockDown: vi.fn(),
  removeBlock: vi.fn(),
  setBlock: vi.fn(),
  showInlineToolbar: vi.fn(),
  moveBlockUp: vi.fn(),
};
