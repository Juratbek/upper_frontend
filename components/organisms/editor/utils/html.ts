import janitor from './janitor';

export function getNodes(html: string): Node[] {
  const wrapper = document.createElement('div');
  wrapper.innerHTML = html;

  return Array.from(wrapper.childNodes);
}

export function isEmpty(element: HTMLElement): boolean {
  const textContent = janitor.clean(element.innerHTML, { onlyText: true });
  return textContent === '';
}
