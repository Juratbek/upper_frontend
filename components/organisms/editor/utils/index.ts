import { EDITOR_BLOCKS_CONTAINER_ID, EDITOR_ROOT_ID } from '../instance/Editor';

export function getEditorBlocksContainer() {
  return document.getElementById(EDITOR_BLOCKS_CONTAINER_ID);
}

export function getEditorRoot() {
  return document.getElementById(EDITOR_ROOT_ID);
}
