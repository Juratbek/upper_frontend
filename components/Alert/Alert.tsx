import { FC } from 'react';
import { getClassName } from 'utils';

import classes from './Alert.module.scss';
import { IAlertProps } from './Alert.types';

export const Alert: FC<IAlertProps> = ({ color = 'yellow', children, ...props }) => {
  const className = getClassName(classes.alert, classes[`alert--${color}`], props.className);

  return <div className={className}>{children}</div>;
};
