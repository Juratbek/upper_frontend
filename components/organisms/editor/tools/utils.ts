import { KeyboardEvent } from 'react';

import { IEditorAPI, IEditorContext } from '../context/EditorContext.types';
import { IBlockData, IBlockNode, TInitialBlockData } from '../instance/Editor.types';
import { isEmpty } from '../utils/html';
import { Selection } from '../utils/selection';

export const generateBlockNode = (element: HTMLDivElement, block: IBlockData): IBlockNode => {
  // just creating new object from params will not work
  // because element is provided using react ref and stores reference to the reference node not to the actual element
  return Object.assign(element, {
    type: block.type,
    id: block.id,
    data: block.data,
  } satisfies IBlockData);
};

export const textBlockKeydownHandler = <T extends { text: string }>(
  event: KeyboardEvent,
  api: IEditorAPI,
  element: HTMLElement,
  blockId: IBlockData['id'],
  config?: { shouldMergeOnBackspace?: boolean },
) => {
  const { code } = event;

  // if user pressed Enter with ctrl or command add new paragraph
  if (code === 'Enter') {
    api.addBlock('paragraph', blockId);
    event.preventDefault();
    return;
  }

  // if paragraph is empty when user presses Back key remove this paragraph
  if (code === 'Backspace' && isEmpty(element)) {
    api
      .removeBlock(blockId)
      .then(({ prevBlocks }) => focusPreviousBlock(prevBlocks, blockId, event));
    return;
  }

  if (code === 'Backspace' && config?.shouldMergeOnBackspace) {
    const selection = Selection.selection;
    if (!selection) return;

    const range = selection.getRangeAt(0);
    const offset = range.startOffset;

    if (offset === 0) {
      api.mergeWithPrevBlock<T>(blockId, (prevBlock, currentBlock) => {
        prevBlock.data.text += ` ${currentBlock.data.text}`;

        return prevBlock.data;
      });
      return;
    }
  }
};

export function getCurrentBlock<T>({ data, hoveredBlock }: IEditorContext<T>) {
  return data.find((b) => hoveredBlock?.id === b.id);
}

function focusPreviousBlock(
  prevBlocks: IBlockData<TInitialBlockData>[],
  blockId: IBlockData['id'],
  event: KeyboardEvent,
) {
  const prevBlock = prevBlocks.find((_, index) => {
    const nextBlock = prevBlocks[index + 1];
    // if next block is the deleted block -> current block is previous block
    if (nextBlock?.id === blockId) return true;

    return false;
  });

  if (!prevBlock) return;

  const prevBlockElement = document.getElementById(prevBlock?.id);
  const textElements = prevBlockElement?.querySelectorAll(
    'h1, h2, h3, h4, h5, h6, p',
  ) as NodeListOf<HTMLParagraphElement | HTMLHeadingElement>;

  if (!textElements || textElements.length < 1) return;

  const lastTextElement = textElements[textElements.length - 1];
  lastTextElement.focus();

  // Create a range (a span of content) and a selection (current user selection)
  const range = document.createRange();
  const selection = window.getSelection();

  if (!selection) return;

  // Move the range to the end of the content inside the element
  range.selectNodeContents(lastTextElement);
  range.collapse(false);

  // Remove any current selections
  selection.removeAllRanges();

  // Add the new range to the selection
  selection.addRange(range);

  event.preventDefault();
}
