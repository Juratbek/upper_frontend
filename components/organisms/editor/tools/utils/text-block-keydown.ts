import { KeyboardEvent } from 'react';

import { IEditorAPI } from '../../context';
import { IBlockData, TInitialBlockData } from '../../instance/Editor.types';
import { isEmpty } from '../../utils/html';
import { Selection } from '../../utils/selection';
import { IParagraphData } from '../paragraph';

const textNodeNames = 'h1, h2, h3, h4, h5, h6, p';

export const textBlockKeydownHandler = <T extends { text: string }>(
  event: KeyboardEvent,
  api: IEditorAPI,
  element: HTMLElement,
  block: Pick<IBlockData, 'id' | 'type'>,
  config?: { shouldMergeOnBackspace?: boolean },
) => {
  const { code } = event;
  const { id: blockId } = block;

  // if user pressed Enter add new paragraph
  if (code === 'Enter') {
    handleEnter(event, api, element, block);
    return;
  }

  // if paragraph is empty when user presses Back key remove this paragraph
  if (code === 'Backspace' && isEmpty(element)) {
    api
      .removeBlock(blockId)
      .then(({ prevBlocks }) => focusPreviousBlock(prevBlocks, blockId, event));
    return;
  }

  if (code === 'Backspace' && config?.shouldMergeOnBackspace) {
    const range = Selection.range;
    if (!range) return;
    const offset = range.startOffset;

    if (offset === 0) {
      api.mergeWithPrevBlock<T>(blockId, (prevBlock, currentBlock) => {
        prevBlock.data.text += ` ${currentBlock.data.text}`;

        return prevBlock.data;
      });
      return;
    }
  }
};

function handleEnter(
  event: KeyboardEvent,
  api: IEditorAPI,
  element: HTMLElement,
  block: Pick<IBlockData, 'id' | 'type'>,
) {
  const { id: blockId, type: blockType } = block;

  const selection = Selection.selection;
  if (!selection) return;

  const range = selection.getRangeAt(0);
  const cursorPosition = range.startOffset;

  // Clone the range so we can modify it without affecting the current selection
  const cloneRange = range.cloneRange();

  // Create a new range that starts from the cursor position to the end of the paragraph
  cloneRange.setStart(range.startContainer, cursorPosition);
  cloneRange.setEndAfter(element.lastChild!);

  // Extract the contents of the range into a DocumentFragment
  const fragment = cloneRange.extractContents();

  // Convert the fragment to HTML
  const temp = document.createElement('div');
  temp.appendChild(fragment);
  const afterCursorHtml = temp.innerHTML;
  const beforeCursorHtml = element.innerHTML;

  // if there is no text after the cursor -> only add a paragraph
  if (!temp.innerText.trim()) {
    api.addBlock('paragraph', blockId);
    event.preventDefault();
    return;
  }

  // update current block state
  api.setBlock<IParagraphData>({
    id: blockId,
    type: blockType,
    data: { text: beforeCursorHtml },
  });

  // update block content (block will not be rendered when text content is updated)
  const currentBlockElement = document.getElementById(blockId)!;
  const textElement = currentBlockElement.querySelector(textNodeNames)!;
  textElement.innerHTML = beforeCursorHtml;

  // add a new block
  api.addBlock('paragraph', blockId, { text: afterCursorHtml }).then((newBlock) => {
    const newBlockElement = document.getElementById(newBlock.id);
    const paragraph = newBlockElement?.querySelector('p');
    paragraph?.focus();
  });

  event.preventDefault();
}

function getLastTextElement(element: HTMLElement) {
  const textElements = element?.querySelectorAll(textNodeNames) as NodeListOf<
    HTMLParagraphElement | HTMLHeadingElement
  >;

  if (!textElements || textElements.length < 1) return null;

  return textElements[textElements.length - 1];
}

function focusPreviousBlock(
  prevBlocks: IBlockData<TInitialBlockData>[],
  blockId: IBlockData['id'],
  event: KeyboardEvent,
) {
  const prevBlock = prevBlocks.find((_, index) => {
    const nextBlock = prevBlocks[index + 1];
    // if next block is the deleted block -> current block is previous block
    if (nextBlock?.id === blockId) return true;

    return false;
  });

  if (!prevBlock) return;

  const prevBlockElement = document.getElementById(prevBlock?.id);

  const lastTextElement = getLastTextElement(prevBlockElement!);
  if (!lastTextElement) return;
  lastTextElement.focus();

  // Create a range (a span of content) and a selection (current user selection)
  const range = document.createRange();
  const selection = window.getSelection();

  if (!selection) return;

  // Move the range to the end of the content inside the element
  range.selectNodeContents(lastTextElement);
  range.collapse(false);

  // Remove any current selections
  selection.removeAllRanges();

  // Add the new range to the selection
  selection.addRange(range);

  event.preventDefault();
}
