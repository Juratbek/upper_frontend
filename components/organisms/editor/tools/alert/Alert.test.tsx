import { fireEvent, render, screen } from '@testing-library/react';
import { vi } from 'vitest';

import { createBlock } from '../../context/EditorContext.utils';
import { mockApi } from '../../mocks';
import { IToolProps } from '../tool.types';
import { Alert } from './Alert.tool';
import { IAlertData } from './Alert.types';

const newData = createBlock('alert', {
  message: 'This is an info message',
  type: 'info',
} satisfies IAlertData);

const mockProps: IToolProps<IAlertData> = {
  id: newData.id,
  type: newData.type,
  data: newData.data as IAlertData,
  isEditable: true,
  api: mockApi,
};

describe('Alert tool', () => {
  beforeEach(() => {
    vi.useFakeTimers();
  });

  afterEach(() => {
    vi.useRealTimers();
  });

  it('snapshot', () => {
    render(<Alert {...mockProps} />);
    expect(document.body).toMatchSnapshot();
  });

  it('snapshot for not editable mode', () => {
    render(<Alert {...mockProps} isEditable={false} />);
    expect(document.body).toMatchSnapshot();
  });

  it('focuses on paragraph on new item created', () => {
    render(<Alert {...mockProps} data={{ type: 'info', message: '' }} />);
    const paragraph = screen.getByRole('paragraph');
    expect(paragraph).toHaveFocus();
  });

  it('debounces the paragraph value', () => {
    const mockSetBlock = vi.fn();
    mockProps.api.setBlock = mockSetBlock;

    render(<Alert {...mockProps} data={{ type: 'info', message: '' }} />);
    const paragraph = screen.getByRole('paragraph');

    fireEvent.input(paragraph, { target: { innerHTML: 'Updated message' } });

    vi.advanceTimersByTime(999);

    expect(mockSetBlock).not.toHaveBeenCalled();
  });

  it('calls the setBlock after 1000 ms', () => {
    const mockSetBlock = vi.fn();
    mockProps.api.setBlock = mockSetBlock;

    render(<Alert {...mockProps} data={{ type: 'info', message: '' }} />);
    const paragraph = screen.getByRole('paragraph');

    fireEvent.input(paragraph, { target: { innerHTML: 'Updated message' } });

    vi.advanceTimersByTime(1000);

    expect(mockSetBlock).toHaveBeenCalledWith({
      id: mockProps.id,
      type: mockProps.type,
      data: { message: 'Updated message', type: mockProps.data.type },
    });
  });
});
