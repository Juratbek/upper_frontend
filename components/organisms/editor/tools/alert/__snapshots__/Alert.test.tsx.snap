// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Alert tool > snapshot 1`] = `
<body>
  <div>
    <div
      class="alert info"
    >
      <span
        class="icon"
      >
        <svg
          fill="none"
          height="24"
          viewBox="0 0 24 24"
          width="24"
        >
          
            
          <path
            d="M9.86674 5.25C10.8149 3.58333 13.1851 3.58333 14.1333 5.25L22.6663 20.25C23.6144 21.9167 22.4293 24 20.533 24H3.46695C1.57072 24 0.385575 21.9167 1.33369 20.25L9.86674 5.25Z"
            fill="#007AFF"
          />
          
            
          <rect
            fill="white"
            height="10"
            width="2"
            x="11"
            y="8"
          />
          
            
          <rect
            fill="white"
            height="2"
            transform="matrix(1 0 0 -1 11 22)"
            width="2"
          />
          
          
        </svg>
      </span>
      <p
        class="text"
        contenteditable="true"
        role="paragraph"
      >
        This is an info message
      </p>
    </div>
  </div>
</body>
`;

exports[`Alert tool > snapshot for not editable mode 1`] = `
<body>
  <div>
    <div
      class="alert info"
    >
      <span
        class="icon"
      >
        <svg
          fill="none"
          height="24"
          viewBox="0 0 24 24"
          width="24"
        >
          
            
          <path
            d="M9.86674 5.25C10.8149 3.58333 13.1851 3.58333 14.1333 5.25L22.6663 20.25C23.6144 21.9167 22.4293 24 20.533 24H3.46695C1.57072 24 0.385575 21.9167 1.33369 20.25L9.86674 5.25Z"
            fill="#007AFF"
          />
          
            
          <rect
            fill="white"
            height="10"
            width="2"
            x="11"
            y="8"
          />
          
            
          <rect
            fill="white"
            height="2"
            transform="matrix(1 0 0 -1 11 22)"
            width="2"
          />
          
          
        </svg>
      </span>
      <p
        class="text"
        contenteditable="false"
        role="paragraph"
      >
        This is an info message
      </p>
    </div>
  </div>
</body>
`;
